File: Src/CSharp/Multiple/Xceed Windows Explorer/Utils/FileSystem/MyComputerFolder.cs
/*
 * Xceed Zip for .NET - Xceed Windows Explorer sample application
 * Copyright (c) 2006 - Xceed Software Inc.
 * 
 * [MyComputerFolder.cs]
 * 
 * Custom implementation of AbstractFolder to represent the "MyComputer" 
 * folder. It basically just returns the list of drives on the machine.
 * 
 * This file is part of Xceed Zip for .NET. The source code in this file 
 * is only intended as a supplement to the documentation, and is provided 
 * "as is", without warranty of any kind, either expressed or implied.
 */

using System;
using System.Collections;
using System.IO;
using Xceed.FileSystem;

namespace Xceed.FileSystem.Samples.Utils.FileSystem
{
  /// <summary>
  /// This class simulate the MyComputer on Windows System. It only expose
  /// all the drives on the machine as child items. 
  /// </summary>
	public class MyComputerFolder : AbstractFolder
	{
    #region CONSTRUCTORS

		public MyComputerFolder()
		{
      this.Refresh();
		}

    #endregion CONSTRUCTORS

    #region PUBLIC PROPERTIES

    protected override bool DoHasAttributes
    {
      get{ return false; }
    }

    protected override System.IO.FileAttributes DoAttributes
    {
      get{ throw new NotSupportedException(); }
      set{ throw new NotSupportedException(); }
    }

    protected override bool DoHasCreationDateTime
    {
      get{ return false; }
    }

    protected override DateTime DoCreationDateTime
    {
      get{ throw new NotSupportedException(); }
      set{ throw new NotSupportedException(); }
    }

    protected override bool DoExists
    {
      get{ return true; }
    }

    protected override string DoFullName
    {
      get{ return m_fullName; }
    }

    protected override bool DoHasLastAccessDateTime
    {
      get{ return false; }
    }

    protected override DateTime DoLastAccessDateTime
    {
      get{ throw new NotSupportedException(); }
      set{ throw new NotSupportedException(); }
    }

    protected override bool DoHasLastWriteDateTime
    {
      get{ return false; }
    }

    protected override DateTime DoLastWriteDateTime
    {
      get{ throw new NotSupportedException(); }
      set{ throw new NotSupportedException(); }
    }

    protected override bool DoIsRoot
    {
      get{ return true; }
    }

    protected override string DoName
    {
      get{ return m_fullName; }
      set{}
    }

    protected override AbstractFolder DoParentFolder
    {
      get{ return null; }
    }

    protected override AbstractFolder DoRootFolder
    {
      get{ return null; }
    }

    #endregion PUBLIC PROPERTIES

    #region PUBLIC METHODS

    protected override void DoCopyTo( FileSystemEventsSession session, FileSystemItem destination, bool replaceExistingFiles )
    {
      throw new NotSupportedException();
    }

    protected override void DoCreate( FileSystemEventsSession session )
    {
      throw new NotSupportedException();
    }

    protected override void DoDelete( FileSystemEventsSession session )
    {
      throw new NotSupportedException();
    }

    protected override FileSystemItem[] DoGetChildItems(FileSystemEventsSession session)
    {
      ArrayList array = new ArrayList( m_children.Count );

      foreach( object folder in m_children )
      {
        // All drives are DiskFolders.
        array.Add( new DiskFolder( folder.ToString() ) );
      }

      return ( FileSystemItem[] )array.ToArray( typeof( FileSystemItem ) );
    }

    protected override AbstractFile DoGetFile( FileSystemEventsSession session, string fileName )
    {
      throw new NotSupportedException();
    }

    protected override AbstractFolder DoGetFolder( FileSystemEventsSession session, string folderName )
    {
      if( m_children.Contains( folderName ) )
        return new DiskFolder( folderName );

      return null;
    }

    protected override void DoRefresh( FileSystemEventsSession session )
    {
      // Clear the actual list.
      m_children.Clear();

      try
      {
        // Get the list of drives on the local machine and add them
        // to the children list.
        string[] drives = Environment.GetLogicalDrives();

        foreach( string drive in drives )
        {
          m_children.Add( drive );
        }
      }
      catch{}
    }

    protected override bool IsPathRooted( string path )
    {
      return false;
    }

    public override bool IsSameAs( FileSystemItem target )
    {
      if( target == null )
        throw new ArgumentNullException( "target" );

      AbstractFolder folder = target as AbstractFolder;

      if( folder == null )
        return false;

      return ( folder.FullName == this.FullName );
    }

    #endregion PUBLIC METHODS

    #region PRIVATE FIELDS

    private const string m_fullName = "My Computer";
    private ArrayList m_children = new ArrayList();

    #endregion PRIVATE FIELDS
	}
}

File: Src/DetailedSamples/Visual Basic .NET/Multiple/Xceed Windows Explorer/Utils/FileSystem/MyComputerFolder.vb
'
 '* Xceed Zip for .NET - Xceed Windows Explorer sample application
 '* Copyright (c) 2006 - Xceed Software Inc.
 '*
'* [MyComputerFolder.vb]
 '*
 '* Custom implementation of AbstractFolder to represent the "MyComputer"
 '* folder. It basically just returns the list of drives on the machine.
 '*
 '* This file is part of Xceed Zip for .NET. The source code in this file
 '* is only intended as a supplement to the documentation, and is provided
 '* "as is", without warranty of any kind, either expressed or implied.
 '


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.IO
Imports Xceed.FileSystem

Namespace Xceed.FileSystem.Samples.Utils.FileSystem
  ''' <summary>
  ''' This class simulate the MyComputer on Windows System. It only expose
  ''' all the drives on the machine as child items. 
  ''' </summary>
  Public Class MyComputerFolder : Inherits AbstractFolder
    #Region "CONSTRUCTORS"

    Public Sub New()
      Me.Refresh()
    End Sub

    #End Region ' CONSTRUCTORS

    #Region "PUBLIC PROPERTIES"

    Protected Overrides ReadOnly Property DoHasAttributes() As Boolean
      Get
        Return False
      End Get
    End Property

    Protected Overrides Property DoAttributes() As System.IO.FileAttributes
      Get
        Throw New NotSupportedException()
      End Get
      Set
        Throw New NotSupportedException()
      End Set
    End Property

    Protected Overrides ReadOnly Property DoHasCreationDateTime() As Boolean
      Get
        Return False
      End Get
    End Property

    Protected Overrides Property DoCreationDateTime() As DateTime
      Get
        Throw New NotSupportedException()
      End Get
      Set
        Throw New NotSupportedException()
      End Set
    End Property

    Protected Overrides ReadOnly Property DoExists() As Boolean
      Get
        Return True
      End Get
    End Property

    Protected Overrides ReadOnly Property DoFullName() As String
      Get
        Return m_fullName
      End Get
    End Property

    Protected Overrides ReadOnly Property DoHasLastAccessDateTime() As Boolean
      Get
        Return False
      End Get
    End Property

    Protected Overrides Property DoLastAccessDateTime() As DateTime
      Get
        Throw New NotSupportedException()
      End Get
      Set
        Throw New NotSupportedException()
      End Set
    End Property

    Protected Overrides ReadOnly Property DoHasLastWriteDateTime() As Boolean
      Get
        Return False
      End Get
    End Property

    Protected Overrides Property DoLastWriteDateTime() As DateTime
      Get
        Throw New NotSupportedException()
      End Get
      Set
        Throw New NotSupportedException()
      End Set
    End Property

    Protected Overrides ReadOnly Property DoIsRoot() As Boolean
      Get
        Return True
      End Get
    End Property

    Protected Overrides Property DoName() As String
      Get
        Return m_fullName
      End Get
      Set
      End Set
    End Property

    Protected Overrides ReadOnly Property DoParentFolder() As AbstractFolder
      Get
        Return Nothing
      End Get
    End Property

    Protected Overrides ReadOnly Property DoRootFolder() As AbstractFolder
      Get
        Return Nothing
      End Get
    End Property

    #End Region ' PUBLIC PROPERTIES

    #Region "PUBLIC METHODS"

    Protected Overrides Sub DoCopyTo(ByVal session As FileSystemEventsSession, ByVal destination As FileSystemItem, ByVal replaceExistingFiles As Boolean)
      Throw New NotSupportedException()
    End Sub

    Protected Overrides Sub DoCreate(ByVal session As FileSystemEventsSession)
      Throw New NotSupportedException()
    End Sub

    Protected Overrides Sub DoDelete(ByVal session As FileSystemEventsSession)
      Throw New NotSupportedException()
    End Sub

    Protected Overrides Function DoGetChildItems(ByVal session As FileSystemEventsSession) As FileSystemItem()
      Dim array As ArrayList = New ArrayList(m_children.Count)

      For Each folder As Object In m_children
        ' All drives are DiskFolders.
        array.Add(New DiskFolder(folder.ToString()))
      Next folder

      Return CType(array.ToArray(GetType(FileSystemItem)), FileSystemItem())
    End Function

    Protected Overrides Function DoGetFile(ByVal session As FileSystemEventsSession, ByVal fileName As String) As AbstractFile
      Throw New NotSupportedException()
    End Function

    Protected Overrides Function DoGetFolder(ByVal session As FileSystemEventsSession, ByVal folderName As String) As AbstractFolder
      If m_children.Contains(folderName) Then
        Return New DiskFolder(folderName)
      End If

      Return Nothing
    End Function

    Protected Overrides Sub DoRefresh(ByVal session As FileSystemEventsSession)
      ' Clear the actual list.
      m_children.Clear()

      Try
        ' Get the list of drives on the local machine and add them
        ' to the children list.
        Dim drives As String() = Environment.GetLogicalDrives()

        For Each drive As String In drives
          m_children.Add(drive)
        Next drive
      Catch
      End Try
    End Sub

    Protected Overrides Function IsPathRooted(ByVal path As String) As Boolean
      Return False
    End Function

    Public Overrides Function IsSameAs(ByVal target As FileSystemItem) As Boolean
      If target Is Nothing Then
        Throw New ArgumentNullException("target")
      End If

      Dim folder As AbstractFolder = CType(IIf(TypeOf target Is AbstractFolder, target, Nothing), AbstractFolder)

      If folder Is Nothing Then
        Return False
      End If

      Return (folder.FullName = Me.FullName)
    End Function

    #End Region ' PUBLIC METHODS

    #Region "PRIVATE FIELDS"

    Private Const m_fullName As String = "My Computer"
    Private m_children As ArrayList = New ArrayList()

    #End Region ' PRIVATE FIELDS
  End Class
End Namespace
